pygerm/reduction.py

```diff
@@ -1,144 +1,297 @@
-    """Bin a single GeRM frame to Energy/ADU vs channel spectrum
-
-    # TODO bin on other channel
-    # TODO make energy range configurable
-    # TODO select chips / channels
-
-    Parameters
-    ----------
-    data : dict
-        must have the keys {'germ_chip', 'germ_chan', 'germ_pd'}
-
-    bins : int or sequence
-       If int, use bins bins between 0 and 70.  If a sequence, is
-       bin edges (including right edge) as with `np.histogram`
-
-    corr_mat : array, optional
-        Correction matrix.  Shaped (2, 12*32).
-        first row is 'm', second row 'b'
-
-    Returns
-    -------
-    spectrum : array
-        This will be shaped (len(bin_edges) - 1, 12*32)
-
-    bin_edges : array
-        The energy / ADU bin edges (including the right most edge)
-
-    """
-    if np.isscalar(bins):
-        if corr_mat is not None:
-            bin_edges = np.linspace(0, 70, bins+1)
-        else:
-            bin_edges = np.linspace(0, 4095, bins+1)
+import matplotlib.pyplot as plt
+from skbeam.core.accumulators.histogram import Histogram
+
+def fix_time(ts, njumps=0, jump=2**29, thresh=2**26):
+    ''' Fix the time in clock cycles from the FPGA
+
+        This is necessary unfortunately when cycling through GeRM data. This
+        takes into account the wrap in time.
+        The time will wrap at jump = 2**bits where bits is the number of bits
+        for the time stamp.
+
+        Normally, this is as straightforward as detecting when the next time
+        stamp wraps back to zero. Unfortunately, the timestamps don't quite
+        come in the right order. Due to this extra complication, we need add a
+        trick where we only correct when we notice a huge difference in times.
+
+        This helps correct for when time deltas are small.
+
+        However, this currently does *not* correct for the one odd case where
+        the time could wrap around, then come back, then wrap around again due
+        to things out of order. This *should* be fixed in the future. It's not
+        clear whether this will be done in hardware or software.
+
+        NOTE :
+            since this depends on time differences, if moving through an
+            array, always give a previous value. For ex:
+            res, njumps = fix_time(arr[:100], njumps=0)
+            res2 = fix_time(arr[100-1:100-1+1], njumps=njumps)
+
+        Parameters
+        ----------
+
+        ts : time (in clock cyles)
+
+        njumps : the previous number of jumps in time
+
+        jump : the max number of clock cycles measured before jump
+            currently 2**29 (29 bits)
+
+        thresh :
+            set some minimum threshold to detect the jump
+            don't want to add a jump at every negative diff
+            since the time stamps don't arrive exactly in order
+            (I was told it's on the order of 12 clock cyles, much less
+            less than my aggressive threhsold of 2**26)
+    '''
+    # if it's dask array compute it
```

Comment:
Would `ts = np.array(ts)` work here?

Reply:
this works thanks! I think we should put `copy=True` just in case (it's `True` by default currently though)